src/Common/tests/System/Collections/ICollection.Generic.Tests.cs

```diff
@@ -0,0 +1,622 @@
+﻿// Licensed to the .NET Foundation under one or more agreements.
+// The .NET Foundation licenses this file to you under the MIT license.
+// See the LICENSE file in the project root for more information.
+
+using System.Collections.Generic;
+using System.Linq;
+using Xunit;
+
+namespace System.Collections.Tests
+{
+    /// <summary>
+    /// Contains tests that ensure the correctness of any class that implements the generic
+    /// ICollection interface
+    /// </summary>
+    public abstract class ICollection_Generic_Tests<T> : IEnumerable_Generic_Tests<T>
+    {
+        #region ICollection<T> Helper Methods
+
+        /// <summary>
+        /// Creates an instance of an ICollection{T} that can be used for testing.
+        /// </summary>
+        /// <returns>An instance of an ICollection{T} that can be used for testing.</returns>
+        protected abstract ICollection<T> GenericICollectionFactory();
+
+        /// <summary>
+        /// Creates an instance of an ICollection{T} that can be used for testing.
+        /// </summary>
+        /// <param name="count">The number of unique items that the returned ICollection{T} contains.</param>
+        /// <returns>An instance of an ICollection{T} that can be used for testing.</returns>
+        protected virtual ICollection<T> GenericICollectionFactory(int count)
+        {
+            ICollection<T> collection = GenericICollectionFactory();
+            AddToCollection(collection, count);
+            return collection;
+        }
+
+        protected virtual bool DuplicateValuesAllowed { get { return true; } }
+        protected virtual bool DefaultValueWhenNotAllowed_Throws { get { return true; } }
+        protected virtual bool IsReadOnly { get { return false; } }
+        protected virtual bool DefaultValueAllowed { get { return true; } }
+        protected virtual IEnumerable<T> InvalidValues { get { return Array.Empty<T>(); } }
+        protected virtual void AddToCollection(ICollection<T> collection, int numberOfItemsToAdd)
+        {
+            int seed = 9600;
+            IEqualityComparer<T> comparer = GetIEqualityComparer();
+            while (collection.Count < numberOfItemsToAdd)
+            {
+                T toAdd = TFactory(seed++);
+                while (collection.Contains(toAdd, comparer) || InvalidValues.Contains(toAdd, comparer))
+                    toAdd = TFactory(seed++);
+                collection.Add(toAdd);
+            }
+        }
+        
+        #endregion
+
+        #region IEnumerable<T> Helper Methods
+
+        protected override IEnumerable<T> GenericIEnumerableFactory(int count)
+        {
+            return GenericICollectionFactory(count);
+        }
+
+        protected override int WaysToModify { get { return 3; } } // add, remove, clear
+
+        protected override void ModifyEnumerable(IEnumerable<T> enumerable, int enumerationCode)
```

Comment:
>Also - I don't have a problem with helper methods in general, but at some point the move towards a helper framework obscures the code being tested too much.

Thanks @Clockwork-Muse, I absolutely agree. I may have overstepped the boundaries of readability in the name of reduced code repetition. I defer to yall on that. I went back and forth on the ModifyEnumerable stuff. I couldn't think of a better system that resolved all of my issues while still being easily understandable. I put in a comment in IEnumerable.Generic.Tests.cs describing the structure but it may be inadequate.

What I wanted to accomplish with ModifyEnumerable and WaysToModify was an inheritable and easily to override system where the modification of an Enumerable (and subsequent enumerator invalidation) could be abstracted out of the tests themselves. The logic around the modification is the same in every case, so I preferred a system where I didn't have to copy-paste all for the tests for every impure method. 

> could you make it an enum rather than code needing to know that 0 == Add, 1 == Remove, etc.?

I wanted the tests for causing an enumerator invalidation to be class-independent, but this raised a huge issue: invalidation comes in many forms. An ICollection can be modified by Add, Remove, or Clear, but a List can also be modified by insert, RemoveAt, sort, etc. This is why I made the admittedly wonky WaysToModify instead of using an enum. From an IEnumerable we don't know the methods that will cause modification of the enumerable so we provide a method for the implementing class to specify them for us.

.>This might be correct, but reading this code is confusing to me... the enumerationCode dictates a removal, but if there's nothing in the collection, we add to it?

It doesn't matter to ModifyEnumerable <i>how</i> the Enumerator is invalidated, only that it is. It is up to the implementing class to determine what to do with the enumerationCode passed to it. In the zero-length case there aren't any objects in the collection so we can't modify it by removal, so we have to modify it some other way instead